flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+						if (c.getData().containsKey(key)) {
+							c.getData().put(key, Base64.getEncoder().encodeToString(serializedStoreHandle));
+						} else {
+							throw new CompletionException(new StateHandleStore.NotExistException(
+								"Could not find " + key + " in ConfigMap " + configMapName));
+						}
+						return Optional.of(c);
+					}
+					return Optional.empty();
+				}).get();
+		} catch (Exception ex) {
+			ExceptionUtils.rethrowException(ExceptionUtils.stripCompletionException(ex));
+		} finally {
+			if (success) {
+				oldStateHandle.discardState();
+			} else {
+				newStateHandle.discardState();
+			}
+		}
+	}
+
+	/**
+	 * Returns the resource version of the ConfigMap or {@link StateHandleStore#NON_EXIST_RESOURCE_VERSION} if the key
+	 * does not exist.
+	 *
+	 * @param key Key in ConfigMap
+	 *
+	 * @return resource version or {@link #NON_EXIST_RESOURCE_VERSION} when not exists.
+	 *
+	 * @throws Exception if the check existence operation failed
+	 */
+	@Override
+	public String exists(String key) throws Exception {
+		checkNotNull(key, "Key in ConfigMap.");
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(configMap -> {
+				if (configMap.getData().containsKey(key)) {
+					return configMap.getResourceVersion();
+				}
+				return StateHandleStore.NON_EXIST_RESOURCE_VERSION;
+			})
+			.orElseThrow(configMapNotExistSupplier);
+	}
+
+	/**
+	 * Gets the {@link RetrievableStateHandle} stored in the given ConfigMap.
+	 *
+	 * @param key Key in ConfigMap
+	 *
+	 * @return The retrieved state handle from the specified ConfigMap and key
+	 *
+	 * @throws IOException if the method failed to deserialize the stored state handle
+	 * @throws NotExistException when the name does not exist
+	 * @throws Exception if get state handle from ConfigMap failed
+	 */
+	@Override
+	public RetrievableStateHandle<T> get(String key) throws Exception {
+		checkNotNull(key, "Key in ConfigMap.");
+
+		final Optional<KubernetesConfigMap> optional = kubeClient.getConfigMap(configMapName);
+		if (optional.isPresent()) {
+			final KubernetesConfigMap configMap = optional.get();
+			if (configMap.getData().containsKey(key)) {
+				return deserializeObject(configMap.getData().get(key));
+			} else {
+				throw new StateHandleStore.NotExistException(
+					"Could not find " + key + " in ConfigMap " + configMapName);
+			}
+		} else {
+			throw configMapNotExistSupplier.get();
+		}
+	}
+
+	/**
+	 * Gets all available state handles sorted by key from Kubernetes.
+	 *
+	 * @return All state handles from ConfigMap.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<Tuple2<RetrievableStateHandle<T>, String>> getAll() {
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(
+				configMap -> {
+					final List<Tuple2<RetrievableStateHandle<T>, String>> stateHandles = new ArrayList<>();
+					configMap.getData().entrySet().stream()
+						.filter(entry -> filter.test(entry.getKey()))
+						.sorted(Comparator.comparing(Map.Entry::getKey))
```

Comment:
Yes. Actually, we do not need to sort the `RetrievableStateHandle` here. Since the user needs to do that if he/she wants to get a specific order. For example, in `ZooKeeperCompletedCheckpointStore#recover`.